File: app/src/main/java/net/wuqs/ontime/db/AlarmDAO.java
package net.wuqs.ontime.db;

import android.arch.lifecycle.LiveData;
import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Delete;
import android.arch.persistence.room.Insert;
import android.arch.persistence.room.OnConflictStrategy;
import android.arch.persistence.room.Query;

import java.util.List;

@Dao
public interface AlarmDAO {
    @Query("SELECT * FROM alarms")
    LiveData<List<AlarmModel>> getAll();

    @Query("SELECT COUNT(*) FROM alarms")
    int count();

    @Query("SELECT * FROM alarms WHERE id IN (:alarmIds)")
    LiveData<List<AlarmModel>> loadAllByIds(int[] alarmIds);

    @Query("SELECT * FROM alarms WHERE id = :alarmId")
    AlarmModel loadAlarm(int alarmId);

    @Query("SELECT * FROM alarms WHERE id = :alarmId")
    LiveData<AlarmModel> getAlarm(int alarmId);

    @Insert
    long[] insertAll(AlarmModel... alarms);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    long insert(AlarmModel alarm);

    @Delete
    void delete(AlarmModel alarm);
}

File: app/src/main/java/net/wuqs/ontime/db/AlarmDao.kt
package net.wuqs.ontime.db

import androidx.lifecycle.LiveData
import androidx.room.*
import androidx.room.OnConflictStrategy.REPLACE

@Dao
interface AlarmDao {

    /** All alarms in the database. */
    @Query("SELECT * FROM alarms ORDER BY next_occurrence IS NULL, next_occurrence")
    fun getAllLive(): LiveData<List<Alarm>>

    /** All alarms in the database. */
    @get:Query("SELECT * FROM alarms")
    val allSync: List<Alarm>

    @get:Query("SELECT * FROM alarms WHERE next_occurrence IS NOT NULL ORDER BY next_occurrence")
    val alarmsHasNextTime: List<Alarm>

    @get:Query("SELECT * FROM alarms WHERE next_occurrence IS NOT NULL ORDER BY next_occurrence")
    val alarmsHasNextTimeLive: LiveData<List<Alarm>>

    @get:Query("SELECT * FROM alarms WHERE historical = 0 ORDER BY next_occurrence")
    val futureAlarmsLive: LiveData<List<Alarm>>

    @get:Query("SELECT * FROM alarms WHERE historical = 1 ORDER BY next_occurrence DESC")
    val historicalAlarmsLive: LiveData<List<Alarm>>

    /** The number of alarms stored in database. */
    @get:Query("SELECT COUNT(*) FROM alarms")
    val count: Int

    /** Returns the [Alarm] of the given id. */
    @Query("SELECT * FROM alarms WHERE id = :id LIMIT 1")
    operator fun get(id: Int): Alarm

    @Query("SELECT * FROM alarms WHERE id = :id LIMIT 1")
    fun getAlarm(id: Int): LiveData<Alarm>

    @Insert(onConflict = REPLACE)
    fun insertAll(vararg alarms: Alarm)

    @Insert(onConflict = REPLACE)
    fun insert(alarm: Alarm)

    @Update
    fun update(alarm: Alarm): Int

    @Update
    fun updateAll(alarms: Collection<Alarm>)

    @Delete
    fun delete(vararg alarms: Alarm): Int
}
